upstairs/src/client.rs

```diff
@@ -0,0 +1,3127 @@
+            log: crucible_common::build_logger(),
+            target_addr: None,
+            repair_addr: None,
+            state: DsState::New,
+            last_flush: JobId(0),
+            stats: DownstairsStats::default(),
+            new_jobs: BTreeSet::new(),
+            skipped_jobs: BTreeSet::new(),
+            region_metadata: None,
+            repair_info: None,
+            io_state_count: ClientIOStateCount::new(),
+        };
+        out.start_dummy_task();
+        out
+    }
+
+    /// Return true if more work can be sent.
+    ///
+    /// If `io_send` can send more work, return true, otherwise return false.
+    /// Because this is used on a `select!` branch that returns
+    /// `futures::future::ready(())`, a short-circuit can be created if no more
+    /// work can actually be done.
+    pub(crate) fn should_do_more_work(&self) -> bool {
+        self.client_task.is_some()
+            && !self.new_jobs.is_empty()
+            && matches!(self.state, DsState::Active | DsState::LiveRepair)
+            && (crate::MAX_ACTIVE_COUNT
+                - self.io_state_count.in_progress as usize)
+                > 0
+    }
+
+    /// Choose which `ClientAction` to apply
+    ///
+    /// This function is called from within a top-level `select!`, so not only
+    /// must the select expressions be cancel safe, but the **bodies** must also
+    /// be cancel-safe.  This is why we simply return a single value in the body
+    /// of each statement.
+    pub(crate) async fn select(&mut self) -> ClientAction {
+        tokio::select! {
+            d = async {
+                if let Some(c) = &mut self.client_task {
+                    tokio::select! {
+                        m = c.client_response_rx.recv() => {
+                            match m {
+                                Some(ClientResponse::Connected) =>
+                                    ClientAction::Connected,
+                                Some(ClientResponse::Message(m)) =>
+                                    ClientAction::Response(m),
+                                Some(ClientResponse::Done(r)) =>
+                                    ClientAction::TaskStopped(r),
+                                None =>
+                                    ClientAction::ChannelClosed,
+                            }
+                        }
+                    }
+                } else {
+                    futures::future::pending().await
+                }
+            } => {
+                d
+            }
+            _ = sleep_until(self.ping_interval), if self.client_task.is_some()
+            => {
+                ClientAction::Ping
+            }
+            _ = sleep_until(self.timeout_deadline) => {
+                ClientAction::Timeout
+            }
+        }
+    }
+
+    /// If the client task is running, send a `Message::HereIAm`
+    ///
+    /// If the client task is **not** running, log a warning to that effect.
```

Comment:
I think this is a matter of phrasing, because there are two almost-but-not-identical things happening.

The client task is **present** if `self.client_task` is `Some(..)`.  This means that we can unwrap the option, but nothing more.

In almost all case, the client task being present also means that it is **running**.  This is because the task waits for its queue to return `None`, which only happens when the `Sender` is dropped (and said sender is stored in `self.client_task`).  This happens [here](https://github.com/oxidecomputer/crucible/blob/9c86e8a6e7b33eb3639fd5faaea89bc14fd31ed9/upstairs/src/client.rs#L2373-L2375).

However, the exception is during program shutdown: Tokio can stop tasks in arbitrary order, so it's possible for `self.client` to be `Some(..)` but for the client task to be stopped.  This is awkward, because it means that something like `client_request_tx` will return an error, since the receiver (in the client task) is dropped!

The equivalent to this case was handled in 51c1f9f6a671b4d5d7249fea8681e325220ad2b7

However, while digging into this, I realized that `client_task` is only `None` in two brief circumstances:
- Upon construction, before an immediate call to `start_task`
- In `Downstairs::reinitialize`, [between `on_missing` (which sets it to `None`) and `reinitialize` (which starts and stores a new task](https://github.com/oxidecomputer/crucible/blob/one-task-to-rule/upstairs/src/downstairs.rs#L713-L740)

In both of these cases, the `None` is transient and we never return to the main `Upstairs::select` while `client_task` is `None`.

As such, we can switch from an `Option<ClientTaskHandle>` to a plain `ClientTaskHandle`, which cleans up some of the code.

This is done in 1743e35337d7a07d5e45241992fbc5112f444bd5, take a look and see if you agree!

Reply:
I like, but now we can remove the "if the client task is not running" comments from these methods as
that path no longer exists.